### ISSUE TITLE
Fix IsModified so that a null equates empty components when using select-before-update.

### ISSUE THREAD

**Issue by mishaurubkov** [nhibernate/nhibernate-core#1486]
The equality check used by the ComponentType.IsModified method must treat empty (all properties null) components equivalent to null component properties and vice versa.

-------------------

Originally this was about another failure in `IsModified`, but checking its current state in Hibernate incites to fix this other more troublesome bug. See this [comment](https://github.com/nhibernate/nhibernate-core/issues/1486#issuecomment-351906653) for all details.

**Comment by hazzik** [comment]
Seems legit. On Hibernate side they "fixed" it in HHH-11237 (https://github.com/hibernate/hibernate-orm/commit/667136c473219ade173b7b3dfea50a1f9304c1a9 + https://github.com/hibernate/hibernate-orm/commit/550c2e38ff41003c178a273b21f038d4295a0b80)

**Comment by mishaurubkov** [comment]
Thank you Alexander, for pointing out. I will follow  HHH-11237 and update the issue.

**Comment by mishaurubkov** [comment]
MOVING ORIGINAL ISSUE TEXT TO THE COMMENTS, REPLACING IT WITH NEW FROM Hibernate HHH-11237
**********************************************************************************************
OLD TITLE:
#1486 ComponentType.IsModified(…) method throws exception (Unable to cast...)
**********************************************************************************************
OLD DESCRIPTION:
**********************************************************************************************
**PROBLEM:**
We are using `EventListener(IPostUpdateEventListener)` in our application to audit data changes:

```c#
foreach (int index in e.Persister.FindModified(e.OldState, e.State, e.Entity, e.Session))
{
    // Log changes
}
```
`FindModified` method calls `IsModified` method for each individual property of the object. `IsModified` throws exception on the component type:

Result StackTrace:
```
at NHibernate.Type.ComponentType.IsModified(Object old, Object current, Boolean[] checkable, ISessionImplementor session) in nhibernate-core-test\src\NHibernate\Type\ComponentType.cs:line 537
```
Result Message:
```
System.InvalidCastException : Unable to cast object of type ‘MyType' to type 'System.Object[]'.
```

**CAUSE:**
NHibernate throws exception in method: 
```c#
ComponentType.IsModified(object old, object current, bool[] checkable, ISessionImplementor session)
```
On the line: 
```c#
object[] oldValues = (Object[]) old;
```
Basically, Nhibernate is trying to cast `ComponentType` to `object[]`. 


**SOLUTION:**
Changing line:
```c#
object[] oldValues = (Object[]) old;
```
TO:
```c#
object[] oldValues = GetPropertyValues(old, session);
```
OR
```c#
object[] oldValues = old is object[] ? (Object[])old : GetPropertyValues(old, session);
```

fixes the problem. 
The method `IsDirty(…)` is done that way: it uses `GetPropertyValues` for “old” and “current” object.

**Comment by hazzik** [comment]
@mishaurubkov have you closed this by accident?

**Comment by mishaurubkov** [comment]
Sorry. Closed by accident.
